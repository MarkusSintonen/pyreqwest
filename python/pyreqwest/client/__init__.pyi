# ruff: noqa: E501, F401
from datetime import timedelta
from types import TracebackType
from typing import Mapping
from typing_extensions import Self

from pyreqwest.http import UrlType
from pyreqwest.middleware import Middleware
from pyreqwest.proxy import Proxy
from pyreqwest.request import RequestBuilder


class Client:
    def request(self, method: str, url: UrlType) -> RequestBuilder: ...
    def get(self, url: UrlType) -> RequestBuilder: ...
    def post(self, url: UrlType) -> RequestBuilder: ...
    def put(self, url: UrlType) -> RequestBuilder: ...
    def patch(self, url: UrlType) -> RequestBuilder: ...
    def delete(self, url: UrlType) -> RequestBuilder: ...
    def head(self, url: UrlType) -> RequestBuilder: ...
    async def __aenter__(self) -> Self: ...
    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None: ...
    async def close(self) -> None: ...


class ClientBuilder:
    def __init__(self) -> None: ...
    def build(self) -> Client: ...
    def with_middleware(self, middleware: Middleware) -> Self: ...
    def max_connections(self, max: int) -> Self: ...
    def error_for_status(self, enable: bool) -> Self: ...
    def user_agent(self, value: str) -> Self: ...
    def default_headers(self, headers: Mapping[str, str]) -> Self: ...
    def cookie_store(self, enable: bool) -> Self: ...
    def gzip(self, enable: bool) -> Self: ...
    def brotli(self, enable: bool) -> Self: ...
    def zstd(self, enable: bool) -> Self: ...
    def deflate(self, enable: bool) -> Self: ...
    def max_redirects(self, max: int) -> Self: ...
    def referer(self, enable: bool) -> Self: ...
    def proxy(self, proxy: Proxy) -> Self: ...
    def no_proxy(self) -> Self: ...
    def timeout(self, timeout: timedelta) -> Self: ...
    def read_timeout(self, timeout: timedelta) -> Self: ...
    def connect_timeout(self, timeout: timedelta) -> Self: ...
    def connection_verbose(self, verbose: bool) -> Self: ...
    def pool_idle_timeout(self, timeout: timedelta) -> Self: ...
    def pool_max_idle_per_host(self, max: int) -> Self: ...
    def http1_title_case_headers(self) -> Self: ...
    def http1_allow_obsolete_multiline_headers_in_responses(self, value: bool) -> Self: ...
    def http1_ignore_invalid_headers_in_responses(self, value: bool) -> Self: ...
    def http1_allow_spaces_after_header_name_in_responses(self, value: bool) -> Self: ...
    def http1_only(self) -> Self: ...
    def http09_responses(self) -> Self: ...
    def http2_prior_knowledge(self) -> Self: ...
    def http2_initial_stream_window_size(self, value: int | None) -> Self: ...
    def http2_initial_connection_window_size(self, value: int | None) -> Self: ...
    def http2_adaptive_window(self, enabled: bool) -> Self: ...
    def http2_max_frame_size(self, value: int | None) -> Self: ...
    def http2_max_header_list_size(self, value: int) -> Self: ...
    def http2_keep_alive_interval(self, value: timedelta | None) -> Self: ...
    def http2_keep_alive_timeout(self, timeout: timedelta) -> Self: ...
    def http2_keep_alive_while_idle(self, enabled: bool) -> Self: ...
    def tcp_nodelay(self, enabled: bool) -> Self: ...
    def local_address(self, addr: str | None) -> Self: ...
    def interface(self, value: str) -> Self: ...
    def tcp_keepalive(self, value: timedelta | None) -> Self: ...
    def tcp_keepalive_interval(self, value: timedelta | None) -> Self: ...
    def tcp_keepalive_retries(self, value: int | None) -> Self: ...
    def tcp_user_timeout(self, timeout: timedelta | None) -> Self: ...
    def add_root_certificate_der(self, cert: bytes) -> Self: ...
    def add_root_certificate_pem(self, cert: bytes) -> Self: ...
    def add_crl_pem(self, cert: bytes) -> Self: ...
    def tls_built_in_root_certs(self, enable: bool) -> Self: ...
    def tls_built_in_webpki_certs(self, enable: bool) -> Self: ...
    def identity_pem(self, buf: bytes) -> Self: ...
    def danger_accept_invalid_hostnames(self, enable: bool) -> Self: ...
    def danger_accept_invalid_certs(self, enable: bool) -> Self: ...
    def tls_sni(self, enable: bool) -> Self: ...
    def min_tls_version(self, value: str) -> Self: ...
    def max_tls_version(self, value: str) -> Self: ...
    def tls_info(self, enable: bool) -> Self: ...
    def https_only(self, enable: bool) -> Self: ...
    def resolve(self, domain: str, ip: str, port: int) -> Self: ...
