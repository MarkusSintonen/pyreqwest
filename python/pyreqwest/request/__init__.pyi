from datetime import timedelta
from typing import Any, Self

from pyreqwest.http import Body, HeaderMap, Url
from pyreqwest.multipart import Form
from pyreqwest.response import Response
from pyreqwest.types import ExtensionsType, FormParams, HeadersType, Middleware, QueryParams, Stream, UrlType

class Request:
    @property
    def method(self) -> str: ...
    @method.setter
    def method(self, value: str) -> None: ...
    @property
    def url(self) -> Url: ...
    @url.setter
    def url(self, value: UrlType) -> None: ...
    @property
    def headers(self) -> HeaderMap: ...
    @headers.setter
    def headers(self, headers: HeadersType) -> None: ...
    @property
    def body(self) -> Body | None: ...
    @body.setter
    def body(self, value: Body | None) -> None: ...
    @property
    def extensions(self) -> dict[str, Any]: ...
    @extensions.setter
    def extensions(self, value: ExtensionsType) -> None: ...
    def copy(self) -> Self: ...
    def __copy__(self) -> Self: ...
    def repr_full(self) -> str: ...
    @classmethod
    def from_request_and_body(cls, request: Request, body: Body | None) -> Self: ...

class ConsumedRequest(Request):
    async def send(self) -> Response: ...

class StreamRequest(Request):
    async def __aenter__(self) -> Response: ...
    async def __aexit__(self, *args: object, **kwargs: Any) -> None: ...
    @property
    def initial_read_size(self) -> int: ...
    @initial_read_size.setter
    def initial_read_size(self, value: int) -> None: ...
    @staticmethod
    def default_initial_read_size() -> int: ...

class RequestBuilder:
    def build_consumed(self) -> ConsumedRequest: ...
    def build_streamed(self) -> StreamRequest: ...
    def error_for_status(self, enable: bool) -> Self: ...
    def with_middleware(self, middleware: Middleware) -> Self: ...
    def header(self, name: str, value: str) -> Self: ...
    def headers(self, headers: HeadersType) -> Self: ...
    def basic_auth(self, username: str, password: str | None) -> Self: ...
    def bearer_auth(self, token: str) -> Self: ...
    def body(self, body: Body | None) -> Self: ...
    def body_bytes(self, body: bytes | bytearray | memoryview) -> Self: ...
    def body_text(self, body: str) -> Self: ...
    def body_json(self, body: Any) -> Self: ...
    def body_stream(self, async_gen: Stream) -> Self: ...
    def timeout(self, timeout: timedelta) -> Self: ...
    def multipart(self, multipart: Form) -> Self: ...
    def query(self, query: QueryParams) -> Self: ...
    def form(self, form: FormParams) -> Self: ...
    def extensions(self, extensions: ExtensionsType) -> Self: ...
