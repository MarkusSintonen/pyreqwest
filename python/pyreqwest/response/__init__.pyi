from typing import Any, Self

from pyreqwest.bytes import Bytes
from pyreqwest.http import HeaderMap, Mime
from pyreqwest.types import ExtensionsType, HeadersType, Stream

class BaseResponse:
    @property
    def status(self) -> int: ...
    @status.setter
    def status(self, value: int) -> None: ...
    @property
    def headers(self) -> HeaderMap: ...
    @headers.setter
    def headers(self, headers: HeadersType) -> None: ...
    @property
    def extensions(self) -> dict[str, Any]: ...
    @extensions.setter
    def extensions(self, value: ExtensionsType) -> None: ...
    @property
    def version(self) -> str: ...
    @version.setter
    def version(self, value: str) -> None: ...
    def error_for_status(self) -> None: ...
    def get_header(self, key: str) -> str | None: ...
    def get_header_all(self, key: str) -> list[str]: ...
    def content_type_mime(self) -> Mime | None: ...

class Response(BaseResponse):
    async def bytes(self) -> Bytes: ...
    async def json(self) -> Any: ...
    async def text(self) -> str: ...
    async def read(self, amount: int = ...) -> Bytes: ...
    @property
    def body_reader(self) -> ResponseBodyReader: ...

class SyncResponse(BaseResponse):
    def bytes(self) -> Bytes: ...
    def json(self) -> Any: ...
    def text(self) -> str: ...
    def read(self, amount: int = ...) -> Bytes: ...
    @property
    def body_reader(self) -> SyncResponseBodyReader: ...

class ResponseBuilder:
    def __init__(self) -> None: ...
    async def build(self) -> Response: ...
    def build_sync(self) -> SyncResponse: ...
    def status(self, value: int) -> Self: ...
    def version(self, value: str) -> Self: ...
    def header(self, name: str, value: str) -> Self: ...
    def headers(self, headers: HeadersType) -> Self: ...
    def extensions(self, extensions: ExtensionsType) -> Self: ...
    def body_bytes(self, body: bytes | bytearray | memoryview) -> Self: ...
    def body_text(self, body: str) -> Self: ...
    def body_json(self, body: Any) -> Self: ...
    def body_stream(self, stream: Stream) -> Self: ...
    def copy(self) -> Self: ...
    def __copy__(self) -> Self: ...

class ResponseBodyReader:
    async def bytes(self) -> Bytes: ...
    async def read(self, amount: int = ...) -> Bytes: ...
    async def read_chunk(self) -> Bytes | None: ...

class SyncResponseBodyReader:
    def bytes(self) -> Bytes: ...
    def read(self, amount: int = ...) -> Bytes: ...
    def read_chunk(self) -> Bytes | None: ...
