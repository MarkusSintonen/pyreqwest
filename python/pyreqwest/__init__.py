"""pyreqwest - Powerful and fast Rust based HTTP client library.

Inspired and using [reqwest](https://github.com/seanmonstar/reqwest).

Feature-rich:
- High performance
- Asynchronous and synchronous HTTP clients
- Built-in support for JSON, form data, and multipart requests
- Connection pooling and keep-alive
- Timeout and retry policies
- HTTPS/TLS support (using [rustls](https://github.com/rustls/rustls))
- Middleware support for advanced request/response processing
- Streaming request and response bodies
- Easy integration with web frameworks
- Cookie management
- Mocking and testing utilities
- Type-safe APIs with Python type hints
- Free threading
"""
