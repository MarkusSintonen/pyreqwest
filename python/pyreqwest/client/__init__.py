from pyreqwest._pyreqwest.client import *
