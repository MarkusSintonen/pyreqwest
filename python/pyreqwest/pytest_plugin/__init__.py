"""PyReqwest pytest plugin for HTTP client mocking."""

from .mock import ClientMocker, RequestMatcher, client_mocker

__all__ = ["ClientMocker", "RequestMatcher", "client_mocker"]
