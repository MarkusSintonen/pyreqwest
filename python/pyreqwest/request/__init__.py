from pyreqwest._pyreqwest.request import *
