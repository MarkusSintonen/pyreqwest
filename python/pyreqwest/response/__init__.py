"""Response classes and builders."""

from pyreqwest._pyreqwest.response import BaseResponse, Response, ResponseBuilder, SyncResponse

__all__ = ["BaseResponse", "Response", "ResponseBuilder", "SyncResponse"]
