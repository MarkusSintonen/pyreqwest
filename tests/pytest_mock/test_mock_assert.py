import json
import re

import pytest
from syrupy import SnapshotAssertion

from pyreqwest.client import ClientBuilder
from pyreqwest.pytest_plugin import ClientMocker
from pyreqwest.request import Request
from pyreqwest.response import ResponseBuilder, Response


async def test_assert_called_default_exactly_once_success(client_mocker: ClientMocker) -> None:
    """Test assert_called with default behavior (exactly once) when mock is called once."""
    mock = client_mocker.get("http://api.example.com/test").with_body_text("response")
    client = ClientBuilder().build()

    await client.get("http://api.example.com/test").build_consumed().send()

    # Should not raise - called exactly once (default)
    mock.assert_called()


async def test_assert_called_default_exactly_once_failure(client_mocker: ClientMocker, snapshot: SnapshotAssertion) -> None:
    """Test assert_called with default behavior when mock is not called."""
    mock = client_mocker.get("http://api.example.com/test").with_body_text("response")
    # Add mock for the unmatched request to prevent real network calls
    client_mocker.get("http://api.example.com/different").with_body_text("different response")
    client = ClientBuilder().build()

    # Make an unmatched request
    await client.get("http://api.example.com/different").build_consumed().send()

    with pytest.raises(AssertionError) as exc_info:
        mock.assert_called()

    assert str(exc_info.value) == snapshot


async def test_assert_called_exact_count_success(client_mocker: ClientMocker) -> None:
    """Test assert_called with exact count when mock is called the right number of times."""
    mock = client_mocker.get("http://api.example.com/test").with_body_text("response")
    client = ClientBuilder().build()

    # Make 3 requests
    for _ in range(3):
        await client.get("http://api.example.com/test").build_consumed().send()

    # Should not raise - called exactly 3 times
    mock.assert_called(count=3)


async def test_assert_called_exact_count_failure(client_mocker: ClientMocker, snapshot: SnapshotAssertion) -> None:
    """Test assert_called with exact count when mock is called wrong number of times."""
    mock = client_mocker.post("http://api.example.com/users") \
        .match_header("Authorization", "Bearer token123") \
        .match_body_json({"name": "John", "age": 30}) \
        .with_status(201) \
        .with_body_json({"id": 1})

    # Add mocks for the unmatched requests
    client_mocker.post("http://api.example.com/users").with_status(400).with_body_text("Bad request")
    client_mocker.get("http://api.example.com/users").with_body_json({"users": []})

    client = ClientBuilder().build()

    # Make one matching request
    await client.post("http://api.example.com/users") \
        .header("Authorization", "Bearer token123") \
        .body_json({"name": "John", "age": 30}) \
        .build_consumed().send()

    # Make some unmatched requests
    await client.post("http://api.example.com/users") \
        .header("Authorization", "Bearer wrong-token") \
        .body_json({"name": "Jane", "age": 25}) \
        .build_consumed().send()

    await client.get("http://api.example.com/users").build_consumed().send()

    with pytest.raises(AssertionError) as exc_info:
        mock.assert_called(count=3)

    assert str(exc_info.value) == snapshot


async def test_assert_called_min_count_success(client_mocker: ClientMocker) -> None:
    """Test assert_called with min_count when mock is called enough times."""
    mock = client_mocker.get("http://api.example.com/test").with_body_text("response")
    client = ClientBuilder().build()

    # Make 5 requests
    for _ in range(5):
        await client.get("http://api.example.com/test").build_consumed().send()

    # Should not raise - called at least 3 times
    mock.assert_called(min_count=3)


async def test_assert_called_min_count_failure(client_mocker: ClientMocker, snapshot: SnapshotAssertion) -> None:
    """Test assert_called with min_count when mock is not called enough times."""
    mock = client_mocker.get("http://api.example.com/endpoint") \
        .match_query({"filter": "active"}) \
        .with_body_json({"data": []})

    # Add mock for the unmatched request
    client_mocker.get("http://api.example.com/endpoint").with_body_json({"data": ["inactive"]})

    client = ClientBuilder().build()

    # Make one matching request
    await client.get("http://api.example.com/endpoint?filter=active").build_consumed().send()

    # Make an unmatched request
    await client.get("http://api.example.com/endpoint?filter=inactive").build_consumed().send()

    with pytest.raises(AssertionError) as exc_info:
        mock.assert_called(min_count=3)

    assert str(exc_info.value) == snapshot


async def test_assert_called_max_count_success(client_mocker: ClientMocker) -> None:
    """Test assert_called with max_count when mock is not called too many times."""
    mock = client_mocker.get("http://api.example.com/test").with_body_text("response")
    client = ClientBuilder().build()

    # Make 2 requests
    for _ in range(2):
        await client.get("http://api.example.com/test").build_consumed().send()

    # Should not raise - called at most 3 times
    mock.assert_called(max_count=3)


async def test_assert_called_max_count_failure(client_mocker: ClientMocker, snapshot: SnapshotAssertion) -> None:
    """Test assert_called with max_count when mock is called too many times."""
    mock = client_mocker.get("http://api.example.com/test").with_body_text("response")
    client = ClientBuilder().build()

    # Make 5 requests
    for _ in range(5):
        await client.get("http://api.example.com/test").build_consumed().send()

    with pytest.raises(AssertionError) as exc_info:
        mock.assert_called(max_count=3)

    assert str(exc_info.value) == snapshot


async def test_assert_called_min_max_range_success(client_mocker: ClientMocker) -> None:
    """Test assert_called with both min_count and max_count when mock is called within range."""
    mock = client_mocker.get("http://api.example.com/test").with_body_text("response")
    client = ClientBuilder().build()

    # Make 3 requests
    for _ in range(3):
        await client.get("http://api.example.com/test").build_consumed().send()

    # Should not raise - called between 2 and 5 times
    mock.assert_called(min_count=2, max_count=5)


async def test_assert_called_min_max_range_failure(client_mocker: ClientMocker, snapshot: SnapshotAssertion) -> None:
    """Test assert_called with both min_count and max_count when mock is called outside range."""
    mock = client_mocker.get("http://api.example.com/test").with_body_text("response")
    client = ClientBuilder().build()

    # Make 1 request (below min_count of 3)
    await client.get("http://api.example.com/test").build_consumed().send()

    with pytest.raises(AssertionError) as exc_info:
        mock.assert_called(min_count=3, max_count=5)

    assert str(exc_info.value) == snapshot


async def test_assert_called_complex_mock_with_all_matchers(client_mocker: ClientMocker, snapshot: SnapshotAssertion) -> None:
    """Test assert_called error message with a complex mock that has all types of matchers."""
    mock = client_mocker.post("http://api.example.com/complex") \
        .match_header("Authorization", re.compile(r"Bearer \w+")) \
        .match_header("Content-Type", "application/json") \
        .match_query({"action": "create", "version": re.compile(r"v\d+")}) \
        .match_body_json({"user": {"name": "John", "role": "admin"}}) \
        .with_status(201)

    # Add mocks for all the unmatched requests
    client_mocker.get("http://api.example.com/complex").with_status(405).with_body_text("Method not allowed")
    client_mocker.post("http://api.example.com/complex").with_status(400).with_body_text("Bad request")

    client = ClientBuilder().build()

    # Make several unmatched requests with different issues
    requests_to_make = [
        # Wrong method
        ("GET", "http://api.example.com/complex?action=create&version=v1", {}, None),
        # Missing auth header
        ("POST", "http://api.example.com/complex?action=create&version=v1", {"Content-Type": "application/json"}, {"user": {"name": "John", "role": "admin"}}),
        # Wrong query param
        ("POST", "http://api.example.com/complex?action=update&version=v1", {"Authorization": "Bearer abc123", "Content-Type": "application/json"}, {"user": {"name": "John", "role": "admin"}}),
        # Wrong body
        ("POST", "http://api.example.com/complex?action=create&version=v1", {"Authorization": "Bearer abc123", "Content-Type": "application/json"}, {"user": {"name": "Jane", "role": "user"}}),
    ]

    for method, url, headers, body in requests_to_make:
        req_builder = getattr(client, method.lower())(url)
        for header_name, header_value in headers.items():
            req_builder = req_builder.header(header_name, header_value)
        if body is not None:
            req_builder = req_builder.body_json(body)
        await req_builder.build_consumed().send()

    with pytest.raises(AssertionError) as exc_info:
        mock.assert_called()

    assert str(exc_info.value) == snapshot


async def test_assert_called_custom_matcher_and_handler(client_mocker: ClientMocker, snapshot: SnapshotAssertion) -> None:
    """Test assert_called error message with custom matcher and handler."""
    async def is_admin_request(request: Request) -> bool:
        if request.body is None:
            return False
        body_bytes = request.body.copy_bytes()
        if body_bytes is None:
            return False
        body_data = json.loads(body_bytes.to_bytes())
        return body_data.get("role") == "admin"

    async def admin_handler(_request: Request) -> Response:
        return await ResponseBuilder.create_for_mocking() \
            .status(200) \
            .body_json({"message": "Admin access granted"}) \
            .build()

    mock = client_mocker.post("http://api.example.com/admin") \
        .match_request(is_admin_request) \
        .match_request_with_response(admin_handler)

    # Add mock for the unmatched request
    client_mocker.post("http://api.example.com/admin").with_status(403).with_body_text("Forbidden")

    client = ClientBuilder().build()

    # Make unmatched request
    await client.post("http://api.example.com/admin") \
        .body_json({"role": "user", "action": "view"}) \
        .build_consumed().send()

    with pytest.raises(AssertionError) as exc_info:
        mock.assert_called(count=2)

    assert str(exc_info.value) == snapshot


async def test_assert_called_with_matched_and_unmatched_requests(client_mocker: ClientMocker, snapshot: SnapshotAssertion) -> None:
    """Test assert_called shows both matched and unmatched requests in error message."""
    mock = client_mocker.get("http://api.example.com/users") \
        .match_query({"active": "true"}) \
        .with_body_json({"users": []})

    # Add mocks for the unmatched requests
    client_mocker.get("http://api.example.com/users").with_body_json({"users": ["inactive"]})
    client_mocker.get("http://api.example.com/posts").with_body_json({"posts": []})

    client = ClientBuilder().build()

    # Make some matching requests
    for i in range(2):
        await client.get(f"http://api.example.com/users?active=true&page={i}").build_consumed().send()

    # Make some unmatched requests
    unmatched_requests = [
        "http://api.example.com/users?active=false",
        "http://api.example.com/users",
        "http://api.example.com/posts?active=true",
    ]

    for url in unmatched_requests:
        await client.get(url).build_consumed().send()

    with pytest.raises(AssertionError) as exc_info:
        mock.assert_called(count=5)  # Expected 5, got 2

    assert str(exc_info.value) == snapshot


async def test_assert_called_many_unmatched_requests_truncation(client_mocker: ClientMocker, snapshot: SnapshotAssertion) -> None:
    """Test that assert_called truncates long lists of unmatched requests."""
    mock = client_mocker.get("http://api.example.com/specific").with_body_text("response")

    # Add mock for the unmatched requests
    client_mocker.get(re.compile(r"http://api\.example\.com/different/.*")).with_body_text("different response")

    client = ClientBuilder().build()

    # Make many unmatched requests (more than the display limit)
    for i in range(8):
        await client.get(f"http://api.example.com/different/{i}") \
            .header("X-Request-ID", f"req-{i}") \
            .build_consumed().send()

    with pytest.raises(AssertionError) as exc_info:
        mock.assert_called()

    assert str(exc_info.value) == snapshot


async def test_assert_called_regex_matchers_display(client_mocker: ClientMocker, snapshot: SnapshotAssertion) -> None:
    """Test assert_called displays regex patterns correctly in error messages."""
    path_pattern = re.compile(r"http://api\.example\.com/users/\d+")
    query_pattern = re.compile(r".*token=\w+.*")
    header_pattern = re.compile(r"Bearer [a-zA-Z0-9]{10,}")
    body_pattern = re.compile(r'.*"action":\s*"(create|update)".*')

    mock = client_mocker.put(path_pattern) \
        .match_query(query_pattern) \
        .match_header("Authorization", header_pattern) \
        .match_body(body_pattern) \
        .with_status(200)

    # Add mock for the unmatched request
    client_mocker.put("http://api.example.com/users/abc").with_status(400).with_body_text("Bad request")

    client = ClientBuilder().build()

    # Make unmatched request
    await client.put("http://api.example.com/users/abc") \
        .header("Authorization", "Bearer short") \
        .body_text('{"action": "delete"}') \
        .build_consumed().send()

    with pytest.raises(AssertionError) as exc_info:
        mock.assert_called()

    assert str(exc_info.value) == snapshot


async def test_assert_called_zero_count_success(client_mocker: ClientMocker) -> None:
    """Test assert_called with count=0 when mock is not called."""
    mock = client_mocker.get("http://api.example.com/test").with_body_text("response")

    # Add mock for the unmatched request
    client_mocker.get("http://api.example.com/different").with_body_text("different response")

    client = ClientBuilder().build()

    # Make unmatched request
    await client.get("http://api.example.com/different").build_consumed().send()

    # Should not raise - called exactly 0 times
    mock.assert_called(count=0)


async def test_assert_called_zero_count_failure(client_mocker: ClientMocker, snapshot: SnapshotAssertion) -> None:
    """Test assert_called with count=0 when mock is called."""
    mock = client_mocker.get("http://api.example.com/test").with_body_text("response")
    client = ClientBuilder().build()

    # Make matching request
    await client.get("http://api.example.com/test").build_consumed().send()

    with pytest.raises(AssertionError) as exc_info:
        mock.assert_called(count=0)

    assert str(exc_info.value) == snapshot

