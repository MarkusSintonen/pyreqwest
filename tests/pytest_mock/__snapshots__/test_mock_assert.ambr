# serializer version: 1
# name: test_assert_called_complex_mock_with_all_matchers
  '''
  Mock was not called as expected.
  Expected exactly 1 call(s), but got 0.
  
  Mock configuration:
    Method: POST
    Path: http://api.example.com/complex
    Query: action=create, version=v\d+ (regex)
    Headers: Authorization: Bearer \w+ (regex), Content-Type: application/json
    Body (JSON): {"user":{"name":"John","role":"admin"}}
  
  Unmatched requests (4):
    1. Request(method='GET', url='http://api.example.com/complex?action=create&version=v1', headers={}, body=None)
    2. Request(method='POST', url='http://api.example.com/complex?action=create&version=v1', headers={'content-type': ['application/json', 'application/json']}, body=BodyBytes(len=39))
    3. Request(method='POST', url='http://api.example.com/complex?action=update&version=v1', headers={'authorization': 'Bearer abc123', 'content-type': ['application/json', 'application/json']}, body=BodyBytes(len=39))
    4. Request(method='POST', url='http://api.example.com/complex?action=create&version=v1', headers={'authorization': 'Bearer abc123', 'content-type': ['application/json', 'application/json']}, body=BodyBytes(len=38))
  '''
# ---
# name: test_assert_called_custom_matcher_and_handler
  '''
  Mock was not called as expected.
  Expected exactly 2 call(s), but got 0.
  
  Mock configuration:
    Method: POST
    Path: http://api.example.com/admin
    Custom matcher: is_admin_request
    Custom handler: admin_handler
  
  Unmatched requests (1):
    1. Request(method='POST', url='http://api.example.com/admin', headers={'content-type': 'application/json'}, body=BodyBytes(len=31))
  '''
# ---
# name: test_assert_called_default_exactly_once_failure
  '''
  Mock was not called as expected.
  Expected exactly 1 call(s), but got 0.
  
  Mock configuration:
    Method: GET
    Path: http://api.example.com/test
  
  Unmatched requests (1):
    1. Request(method='GET', url='http://api.example.com/different', headers={}, body=None)
  '''
# ---
# name: test_assert_called_exact_count_failure
  '''
  Mock was not called as expected.
  Expected exactly 3 call(s), but got 1.
  
  Mock configuration:
    Method: POST
    Path: http://api.example.com/users
    Headers: Authorization: Bearer token123
    Body (JSON): {"name":"John","age":30}
  
  Unmatched requests (2):
    1. Request(method='POST', url='http://api.example.com/users', headers={'authorization': 'Bearer wrong-token', 'content-type': 'application/json'}, body=BodyBytes(len=24))
    2. Request(method='GET', url='http://api.example.com/users', headers={}, body=None)
  
  Matched requests (1):
    1. Request(method='POST', url='http://api.example.com/users', headers={'authorization': 'Bearer token123', 'content-type': 'application/json'}, body=BodyBytes(len=24))
  '''
# ---
# name: test_assert_called_many_unmatched_requests_truncation
  '''
  Mock was not called as expected.
  Expected exactly 1 call(s), but got 0.
  
  Mock configuration:
    Method: GET
    Path: http://api.example.com/specific
  
  Unmatched requests (8):
    1. Request(method='GET', url='http://api.example.com/different/3', headers={'x-request-id': 'req-3'}, body=None)
    2. Request(method='GET', url='http://api.example.com/different/4', headers={'x-request-id': 'req-4'}, body=None)
    3. Request(method='GET', url='http://api.example.com/different/5', headers={'x-request-id': 'req-5'}, body=None)
    4. Request(method='GET', url='http://api.example.com/different/6', headers={'x-request-id': 'req-6'}, body=None)
    5. Request(method='GET', url='http://api.example.com/different/7', headers={'x-request-id': 'req-7'}, body=None)
    ... and 3 more
  '''
# ---
# name: test_assert_called_max_count_failure
  '''
  Mock was not called as expected.
  Expected at most 3 call(s), but got 5.
  
  Mock configuration:
    Method: GET
    Path: http://api.example.com/test
  
  Matched requests (5):
    1. Request(method='GET', url='http://api.example.com/test', headers={}, body=None)
    2. Request(method='GET', url='http://api.example.com/test', headers={}, body=None)
    3. Request(method='GET', url='http://api.example.com/test', headers={}, body=None)
    ... and 2 more
  '''
# ---
# name: test_assert_called_min_count_failure
  '''
  Mock was not called as expected.
  Expected at least 3 call(s), but got 1.
  
  Mock configuration:
    Method: GET
    Path: http://api.example.com/endpoint
    Query: filter=active
  
  Unmatched requests (1):
    1. Request(method='GET', url='http://api.example.com/endpoint?filter=inactive', headers={}, body=None)
  
  Matched requests (1):
    1. Request(method='GET', url='http://api.example.com/endpoint?filter=active', headers={}, body=None)
  '''
# ---
# name: test_assert_called_min_max_range_failure
  '''
  Mock was not called as expected.
  Expected at least 3 and at most 5 call(s), but got 1.
  
  Mock configuration:
    Method: GET
    Path: http://api.example.com/test
  
  Matched requests (1):
    1. Request(method='GET', url='http://api.example.com/test', headers={}, body=None)
  '''
# ---
# name: test_assert_called_regex_matchers_display
  '''
  Mock was not called as expected.
  Expected exactly 1 call(s), but got 0.
  
  Mock configuration:
    Method: PUT
    Path: http://api\.example\.com/users/\d+ (regex)
    Query: .*token=\w+.* (regex)
    Headers: Authorization: Bearer [a-zA-Z0-9]{10,} (regex)
    Body (text): .*"action":\s*"(create|update)".* (regex)
  
  Unmatched requests (1):
    1. Request(method='PUT', url='http://api.example.com/users/abc', headers={'authorization': 'Bearer short'}, body=BodyBytes(len=20))
  '''
# ---
# name: test_assert_called_with_matched_and_unmatched_requests
  '''
  Mock was not called as expected.
  Expected exactly 5 call(s), but got 2.
  
  Mock configuration:
    Method: GET
    Path: http://api.example.com/users
    Query: active=true
  
  Unmatched requests (3):
    1. Request(method='GET', url='http://api.example.com/users?active=false', headers={}, body=None)
    2. Request(method='GET', url='http://api.example.com/users', headers={}, body=None)
    3. Request(method='GET', url='http://api.example.com/posts?active=true', headers={}, body=None)
  
  Matched requests (2):
    1. Request(method='GET', url='http://api.example.com/users?active=true&page=0', headers={}, body=None)
    2. Request(method='GET', url='http://api.example.com/users?active=true&page=1', headers={}, body=None)
  '''
# ---
# name: test_assert_called_zero_count_failure
  '''
  Mock was not called as expected.
  Expected exactly 0 call(s), but got 1.
  
  Mock configuration:
    Method: GET
    Path: http://api.example.com/test
  
  Matched requests (1):
    1. Request(method='GET', url='http://api.example.com/test', headers={}, body=None)
  '''
# ---
